frontend/src/scenes/experiments/Metrics/ExperimentMetricModal.tsx

```diff
@@ -23,37 +23,62 @@ export function ExperimentMetricModal({
         editingPrimaryMetricIndex,
         editingSecondaryMetricIndex,
     } = useValues(experimentLogic({ experimentId }))
-    const {
-        setMetric,
-        updateExperimentMetrics,
-        setExperiment,
-        closePrimaryMetricModal,
-        closeSecondaryMetricModal,
-        restoreUnmodifiedExperiment,
-    } = useActions(experimentLogic({ experimentId }))
+    const { setMetric, updateExperimentMetrics, setExperiment, closePrimaryMetricModal, closeSecondaryMetricModal } =
+        useActions(experimentLogic({ experimentId }))
 
     const metricIdx = isSecondary ? editingSecondaryMetricIndex : editingPrimaryMetricIndex
     const metricsField = isSecondary ? 'metrics_secondary' : 'metrics'
 
-    const handleSetMetric = useCallback(
-        (newMetric: ExperimentMetric): void => {
-            if (metricIdx == null) {
+    const metrics = experiment[metricsField]
+    const originalMetric = metricIdx != null ? (metrics[metricIdx] as ExperimentMetric) : null
+
+    // Local state for the entire metric - only updates global state on save
+    const [localMetric, setLocalMetric] = useState<ExperimentMetric | null>(originalMetric)
+
+    // Update local state when metric changes (e.g., switching between metrics)
+    useEffect(() => {
+        setLocalMetric(originalMetric)
+    }, [originalMetric])
```

Comment:
Using `useEffect` here is correct, although we should always try to avoid it. In this case, one way will be to make `originalMetric` a prop and remount the dialog when we need it.
I've been looking at our modals, and we usually mount them when the page is rendered... so this is cool, but we need a larger refactor of our modals to either mount them on demand, or mount a `Content` component based on visibility.

Reply:
Would you say we should just not do the optimization for now and leave it as is to avoid using `useEffect`?